### PR TITLE
#11144: adding cache invalidation for classification store KeyConfig

### PR DESCRIPTION
**Pull request by awroblewski**
Pull request as per issue #11144

@brusch , as suggested I've created a static method to separate the cache key creation logic as per your given example. The cache keys for the classification store keys are now calculated using this method.

@weisswurstkanone , I've implemented the same for the GroupConfig, however, the GroupConfig did not use caching on behalf of the group name but just the id. So the current cache invalidation stays the same. Does it probably make sense to enable the same caching on behalf of the group name as it is done with the KeyConfig or was it left out on purpose?

Thank you

### CODE REVIEW

**Comment by brusch** [comment]
@awroblewski you are right, I think it also would make sense to do the "named" based caching as well for `GroupConfig`.  Could you implement this as well in this PR? Thanks a lot!

File: models/DataObject/Classificationstore/KeyConfig.php [hunk 1]
```diff
@@ -424,16 +396,60 @@ public function setStoreId($storeId)
...
+
+        return Cache::load($cacheKey);
+    }
+
+    /**
+     * Calculate cache key
+     *
      * @param int $id
-     * @param string|null $name
+     * @param string $name
```
Comment (blankse) [inline models/DataObject/Classificationstore/KeyConfig.php:383]:
```suggestion
     * @param string|null $name
```

**Comment by awroblewski** [comment]
Hello @brusch 

I have implemented caching for the group as well expanded the same onto the collections.

I got one more thing. I have again modified the KeyConfig, as there was an additonal bug I discovered during my tests.

The KeyConfig.php implements another runtime caching (next to the Pimcore Runtime Caching) in the KeyConfig class using a local array which was toggled using the variable $cacheenabled. Whenever you request a KeyConfig using an id and have $cacheenbaled on true the if clause tries to retrieve a value from the array by the given index $id. This fails with an exception when the requested $id was not stored yet. Example:

`if(self::$cacheEnabled && self::$cache[$id]) {...}`

If I request a KeyConfig with id 29 which is not yet stored in the array with cacheenabled = true it yíelds an exception

`"No entry is registered for key '29'"`

I'm not sure why the KeyConfig implements another local array caching in the first place while at the same time using the Runtime Cache. It is practically the same, right?

I have remove this local array caching for now, however, if i miss something here I can put that back in, let me know.

As well I have consolidated the cache getting, setting and removing into static functions for Key-,Group-, and CollectionConfig. This made the implementation more consisten across the three scopes.


Thank You :)

File: models/DataObject/Classificationstore/KeyConfig.php (hunk 1 above)
Comment (awroblewski) [reply models/DataObject/Classificationstore/KeyConfig.php:383]: @blankse 

thanks for the catch :)

File: models/DataObject/Classificationstore/CollectionConfig.php [hunk 2]
```diff
@@ -262,4 +323,64 @@ public function setStoreId($storeId)
...
+     * @return CollectionConfig|bool
+     */
+    private static function getCache(string $cacheKey): CollectionConfig|bool
+    {
+        if (self::$cacheEnabled && Cache\Runtime::isRegistered($cacheKey) && $config = Cache\Runtime::get($cacheKey)) {
+            return $config;
+        }
+
+        return Cache::load($cacheKey);
+    }
```
Comment (brusch) [inline models/DataObject/Classificationstore/CollectionConfig.php]: To avoid code duplication and make it easier to maintain, we should put that code into a trait. Since this is very generic, i think we can put this trait here: `lib/Cache/RuntimeCacheTrait.php`. 
As a follow up we could then also migrate existing usages of the runtime cache to this trait.

File: models/DataObject/Classificationstore/CollectionConfig.php [hunk 3]
```diff
@@ -108,6 +131,22 @@ public static function create()
...
+
+    /**
+     * @return bool
+     */
+    public static function getCacheEnabled()
+    {
+        return self::$cacheEnabled;
+    }
+
```
Comment (brusch) [inline models/DataObject/Classificationstore/CollectionConfig.php]: This should go to the trait as well.

File: models/DataObject/Classificationstore/CollectionConfig.php [hunk 4]
```diff
@@ -25,13 +26,18 @@
...
     /**
-     * Group id.
-     *
-     * @var int
+     * @var bool
+     */
+    public static $cacheEnabled = false;
+
+    /**
+     * @var int|null
```
Comment (brusch) [inline models/DataObject/Classificationstore/CollectionConfig.php]: Also a candidate for the trait and additionally it could be `private`.

File: models/DataObject/Classificationstore/CollectionConfig.php [hunk 5]
```diff
@@ -247,6 +294,20 @@ public function getCreationDate()
...
+     */
+    public function getRelations()
+    {
+        $list = new CollectionGroupRelation\Listing();
+        $list->setCondition('colId = ' . $this->id);
+        $list = $list->load();
+
+        return $list;
+    }
+
```
Comment (brusch) [inline models/DataObject/Classificationstore/CollectionConfig.php]: This code is not really related to this PR, so we should put it into another PR targeting `10.x`.

**Review by brusch — CHANGES_REQUESTED** [review]
Thanks a lot, looks already pretty good, however there are just some more small optimizations 😉 Could you update the PR until noon - that would be nice, then we could include it into the `10.2.10` release. Thanks in advance!

File: models/DataObject/Classificationstore/CollectionConfig.php (hunk 2 above)
Comment (awroblewski) [reply models/DataObject/Classificationstore/CollectionConfig.php]: Done...

File: models/DataObject/Classificationstore/CollectionConfig.php (hunk 3 above)
Comment (awroblewski) [reply models/DataObject/Classificationstore/CollectionConfig.php]: Done...

File: models/DataObject/Classificationstore/CollectionConfig.php (hunk 4 above)
Comment (awroblewski) [reply models/DataObject/Classificationstore/CollectionConfig.php]: Done...

File: models/DataObject/Classificationstore/CollectionConfig.php (hunk 5 above)
Comment (awroblewski) [reply models/DataObject/Classificationstore/CollectionConfig.php]: You're right I will raise another PR for this. Removed...

File: lib/Cache/RuntimeCacheTrait.php [hunk 6]
```diff
@@ -0,0 +1,83 @@
+<?php
+/**
+ *
+ * Created by PhpStorm.
+ *
+ * @date: 02.02.2022
+ *
+ * @author: Andreas Wroblewski <andreas.wroblewski@codafish.net>
+ * @copyright: codafish><> GmbH, https://codafish.net
+ */
```
Comment (brusch) [inline lib/Cache/RuntimeCacheTrait.php]:
```suggestion
/**
 * Pimcore
 *
 * This source file is available under two different licenses:
 * - GNU General Public License version 3 (GPLv3)
 * - Pimcore Commercial License (PCL)
 * Full copyright and license information is available in
 * LICENSE.md which is distributed with this source code.
 *
 *  @copyright  Copyright (c) Pimcore GmbH (http://www.pimcore.org)
 *  @license    http://www.pimcore.org/license     GPLv3 and PCL
 */
```

**Review by brusch — CHANGES_REQUESTED** [review]

**Comment by awroblewski** [comment]
@brusch I have refactored the classes. They are now using the Trait in lib/Cache/RuntimeCacheTrait.php.

I was not sure how to handle the Trait signature. Currently i left the standard signature which applies to myself but shouldn't be your Pimcore signature in here? Let me know if I have to change this.

many thx

**Comment by brusch** [comment]
@awroblewski thanks a lot, looks good to me now. The license header should be the one from Pimcore, see code suggestion. Thanks!

**Comment by awroblewski** [comment]
@brusch , just have seen this while writing the comment, all good now
